Allwinner: Support for DT overlays

jernejsk opened the pull request:
This PR is initial work to support DT overlays also for other than RPi ARM SoCs. Several users already asked for that functionality directly or indirectly (how to modify dtb file?).

Note that this mostly just brings up infrastructure with one simple Allwinner H6 overlay. Please tell me if you concur with approach taken here and what would you change.

Before this is merged I would expect that dt-overlay repo is moved to LE github and maintained there.

dhewg commented:
Nice, haven't tested it, but LGTM!

jernejsk commented:
It's not in shape to merged just yet. There are also few things to consider, like which license to use in DT fragments. I vote for standard DT license `SPDX-License-Identifier: (GPL-2.0+ or MIT)`.

dhewg commented:
Yeah, that sounds sensible.

I don't foresee too many overlays, and they're all tiny. On top of that I guess some may need minor changes if the upstream dts changes, so in a way vaguely bound to the kernel version. So why not just add them to packages/tools/dt-overlays? That'll make it a little easier to maintain

jernejsk commented:
DT rarely changes considerably, so I don't think we need to develop mechanism for differentiating between kernel versions. Besides, that is important only when stable is branched. In such cases, we can maintain two branches in dt-overlay repo.

I'm ok with including dt-overlay sources in build system directly, but historical way of doing such things it to make separate repo with all the files.

BTW, I think we should add README files for each fragment separately and script, which filters useful fragments per platforms should concatenate them in one common README file, so user would have example when and how to use them. This would be similar how RPi functions and how Armbian deals with DT overlays.

@CvH @chewitt @Kwiboo  any thoughts?

clementperon commented on packages/tools/u-boot/patches/pxe-add-support-for-FDT-overlays.patch:
```diff
@@ -0,0 +1,197 @@
++		return;
++
++	/* Get the specific overlay loading address */
++	fdtoverlay_addr_env = env_get("fdtoverlay_addr_r");
```
We need to declare this in the bootloader env

jernejsk commented:
This PR works now properly - runtime tested. I'll add some kind of README handling before I'll mark it as ready.

jernejsk commented:
I think it's better to write a article about overlays on wiki and just select meaningful names for overlays so users could quickly find wanted overlay.

Example:
Enabling IR on Pine64 Plus is as simple as adding following line to extlinux.conf:
`FDTOVERLAYS /overlays/sun50i-a64-ir.dtbo`

If wifi+bt module is also connected to the board, following line would be added:
`FDTOVERLAYS /overlays/sun50i-a64-ir.dtbo /overlays/sun50i-a64-pine64-wifi-bt.dtbo`

For overlay sources I decided to use dual licence GPLv2+ or MIT, which is commonly used for DT files in Linux.

@mklein-de Can you test this PR on BPi Zero? I added sun8i-h2-plus-bpi-m2-zero-ethernet.dtbo overlay for ethernet and I would like to know if it works. I suggest you build full image from scratch (U-Boot must be rebuild and updated).

@HiassofT can you please check if linux package change affects RPi overlays? It shouldn't, since RPi kernel enables same flag, but you never know...

HiassofT commented:
DTC_FLAGS change looks fine, I'll try to do a test the next couple of days and report back

HiassofT commented:
Quick RPi4 build and runtime test with this PR was fine.

One thing I'm wondering is if we should maybe namespace the overlays into project specific dirs (eg /flash/overlay/allwinner/ etc).

This could make things easier when more projects follow adding DT support and when we eventually create unified ARMv7/v8 projects - then overlays would already be nicely separated

mklein-de commented:
> @mklein-de Can you test this PR on BPi Zero? I added sun8i-h2-plus-bpi-m2-zero-ethernet.dtbo overlay for ethernet and I would like to know if it works. I suggest you build full image from scratch (U-Boot must be rebuild and updated).

Yes, ethernet overlay works fine with a fresh BPi Zero image here.

The only minor annoyance is that I get a new (random) MAC address with each boot. Armbian has a [u-boot patch](https://github.com/armbian/build/blob/master/patch/u-boot/u-boot-sunxi/0020-sunxi-call-fdt_fixup_ethernet-again-to-set-macaddr-f.patch) to remedy this. Ok to open a new PR?

jernejsk commented:
Let me check if I can come up with patch which can be upstreamed to U-Boot right now. That one looks almost fine otherwise.

jernejsk commented:
@mklein-de go ahead with PR, just put patch to common folder, e.g. projects/Allwinner/patches/u-boot - that way this will be used for all devices and it will allow to drop U-Boot patch for H6...

jernejsk commented:
Repo moved under LE github.

@HiassofT I wouldn't complicate things at the beginning. Unified images are still some time off and I highly doubt direct update will be possible.

jernejsk commented:
Rebased, gtg or is there something to fix?
